src/Cli/Microsoft.DotNet.Cli.Utils/TypoCorrection.cs

```diff
@@ -7,40 +7,101 @@
+        /// max number of suggestion is restricted to 10 entries
         /// </summary>
         /// <param name="possibleTokens">List of tokens to select from.</param>
         /// <param name="currentToken">The token that is being compared.</param>
         /// <param name="maxLevenshteinDistance">the difference between two strings, default: 3.</param>
         /// <returns>The enumerator to tokens similar to <paramref name="currentToken"/>.</returns>
         public static IEnumerable<string> GetSimilarTokens(IEnumerable<string> possibleTokens, string currentToken, int maxLevenshteinDistance = 3)
         {
-            int? bestDistance = null;
-            return possibleTokens
-                .Select(possibleMatch => (possibleMatch, distance: GetDistance(currentToken, possibleMatch)))
-                .Where(tuple => tuple.distance <= maxLevenshteinDistance)
-                .OrderBy(tuple => tuple.distance)
-                .ThenByDescending(tuple => GetStartsWithDistance(currentToken, tuple.possibleMatch))
-                .TakeWhile(tuple =>
+            var minCurrentTokenLength = 3;
+            var maxNumberOfSuggestions = 10;
+
+            var numberOfSuggestions = 0;
+            var currentTokenLength = currentToken.Length;
+            var possibleSuggestions = possibleTokens.Select((string possibleMatch) => new Suggestion(possibleMatch, possibleMatch.Length));
+
+            var matchByStartsWith = possibleSuggestions
+               .Where(s => s.PossibleMatch.StartsWith(currentToken))
+               .OrderBy(s => s.Distance)
+               .ToList();
+
+            numberOfSuggestions += matchByStartsWith.Count;
+            if (numberOfSuggestions >= maxNumberOfSuggestions)
+            {
+                return matchByStartsWith
+                    .Take(maxNumberOfSuggestions)
+                    .Select(s => s.PossibleMatch);
+            }
+
+            var matchByContains = new List<Suggestion>();
+            if (currentToken.Length >= minCurrentTokenLength)
+            {
+                matchByContains = possibleSuggestions
+                    .Where(s => !matchByStartsWith.Contains(s)
+                        && s.PossibleMatch.Contains(currentToken)
+                        && s.Distance - currentTokenLength <= maxLevenshteinDistance)
+                    .OrderBy(s => s.Distance)
+                    .ToList();
+
+                numberOfSuggestions += matchByContains.Count;
+                if (numberOfSuggestions >= maxNumberOfSuggestions)
                 {
-                    (string _, int distance) = tuple;
-                    bestDistance ??= distance;
-                    return distance == bestDistance;
-                })
-                .Select(tuple => tuple.possibleMatch);
+                    return matchByStartsWith
+                        .Concat(matchByContains)
+                        .Take(maxNumberOfSuggestions)
```

Comment:
Perf suggestion: Similar to above, move the `Take` to the previous statement. `Take(maxNumberOfSuggestions - numberOfSuggestions)`